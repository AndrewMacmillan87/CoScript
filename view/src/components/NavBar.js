import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";
import Button from "./buttons/Button";

import { getUserMessages } from "../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./navbar.css";

export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            visible: false,
            unread: false,
        };
        this.handleClick = this.handleClick.bind(this);
    }

    /**
     * Checks to see if the user has any unread messages
     * in their personal profile message inbox and sets
     * the 'unread' state property to true if they have.
     * This displays a small icon to the user to indicate
     * they have unread messages
     */
    async componentDidMount() {
        const userMessages = await getUserMessages();
        for (let message of userMessages) {
            if (!message.read) {
                this.setState({
                    unread: true,
                });
            }
        }
    }

    /**
     * Sets the 'visible' state property so that the
     * dropdown menu can be revealed and hidden by
     * clicking the link
     */
    handleClick() {
        this.setState((state) => ({
            visible: !state.visible,
        }));
    }

    render() {
        const { unread } = this.state;
        const { user, loading, history } = this.props;

        const menuIcon = require("../img/menu.svg");
        const spinner = require("../img/loading.svg");
        const isLoggedIn = user && user.isLoggedIn | false;

        const loggedIn = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                        <Link
                            to='#'
                            onClick={async () => {
                                await this.props.getMostRecentProject();
                                const { projectData } = this.props;
                                projectData
                                    ? history.push(`/editor${projectData.url}`)
                                    : history.push("/editor");
                            }}>
                            Workspace
                        </Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    {user && !loading ? (
                        <div className='logged-in-user-container'>
                            <p className='user-username'>
                                {user.profile.displayName}
                            </p>
                            <button
                                className='desktop-icon-link'
                                onClick={this.handleClick}>
                                {unread ? (
                                    <div className='avatar-unread-msg-container'>
                                        <div className='avatar-unread-msg-icon'></div>
                                        <img
                                            className='user-gravatar'
                                            src={user.avatar}
                                            alt='user gravatar'
                                        />
                                    </div>
                                ) : (
                                    <img
                                        className='user-gravatar'
                                        src={user.avatar}
                                        alt='user gravatar'
                                    />
                                )}
                            </button>
                        </div>
                    ) : (
                        <img
                            className='user-spinner'
                            src={spinner}
                            alt='loading spinner icon'
                        />
                    )}
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                    <Link
                        to='#'
                        onClick={async () => {
                            await this.props.getMostRecentProject();
                            const { projectData } = this.props;
                            projectData
                                ? history.push(`/editor${projectData.url}`)
                                : history.push("/editor");
                        }}>
                        <p className='dropdown-link'>Workspace</p>
                    </Link>
                    <Link to='/profile'>
                        <p className='dropdown-link'>My Profile</p>
                    </Link>
                    <a
                        onClick={() => {
                            history["from"] = "logout";
                            history.push("/redirect");
                        }}
                        href=' '>
                        <p className='dropdown-link'>Logout</p>
                    </a>
                </div>
            </Fragment>
        );

        const visiting = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    <div>
                        <Link to='/register'>
                            <Button classes='btn' value='Register' />
                        </Link>
                    </div>
                    <div>
                        <Link to='/login'>
                            <Button classes='grn-btn' value='Login' />
                        </Link>
                    </div>
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/register'>
                        <p className='dropdown-link'>Register</p>
                    </Link>
                    <Link to='/login'>
                        <p className='dropdown-link'>Login</p>
                    </Link>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                </div>
            </Fragment>
        );

        return (
            <nav className='navbar'>
                {!loading && (
                    <Fragment>{!isLoggedIn ? visiting : loggedIn}</Fragment>
                )}
            </nav>
        );
    }
}

NavBar.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(setProps, {
    getMostRecentProject,
})(withRouter(NavBar));
