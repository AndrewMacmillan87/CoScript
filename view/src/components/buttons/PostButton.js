import React from "react";
import PropTypes from "prop-types";

import "./buttons.css";

const PostButton = ({ onClick, classes, value }) => {
    return (
        <input
            onClick={onClick}
            className={classes}
            type='submit'
            value={value}
        />
    );
};

PostButton.defaultProps = {
    value: "Verify"
};

PostButton.propTypes = {
    onClick: PropTypes.func,
    classes: PropTypes.string,
    value: PropTypes.string
};

export default PostButton;
