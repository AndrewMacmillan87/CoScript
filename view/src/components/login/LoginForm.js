import React from "react";
import { Link, Redirect } from "react-router-dom";
import PostButton from "../buttons/PostButton";
import PropTypes from "prop-types";

import Alert from "../alert/Alert";

const LoginForm = ({
    username,
    password,
    loading,
    onChange,
    onSubmit,
    isLoggedIn,
    usernameField
}) => {
    if (isLoggedIn) {
        return <Redirect to='/' />;
    }

    const spinner = require("../../img/loading.svg");

    return (
        <section className='login-cover'>
            <div className='login-container'>
                <h1 className='login-header'>Login...</h1>
                <div>
                    <div className='login-form'>
                        <form onSubmit={e => onSubmit(e)}>
                            <label htmlFor='username' className='log-label'>
                                Username:
                            </label>
                            <input
                                className='log-input'
                                type='text'
                                name='username'
                                value={username}
                                onChange={e => onChange(e)}
                                placeholder='Username...'
                                onFocus={e => (e.target.placeholder = "")}
                                onBlur={e =>
                                    (e.target.placeholder = "Username...")
                                }
                                ref={usernameField}
                            />
                            <br />
                            <label htmlFor='password' className='log-label'>
                                Password:
                            </label>
                            <input
                                className='log-input'
                                type='password'
                                name='password'
                                value={password}
                                onChange={e => onChange(e)}
                                placeholder='Password...'
                                onFocus={e => (e.target.placeholder = "")}
                                onBlur={e =>
                                    (e.target.placeholder = "Password...")
                                }
                            />
                            <br />
                            <p className='log-link'>
                                Haven't registered? You can do that{" "}
                                <Link to='/register'>
                                    <span className='login-link'>here</span>
                                </Link>
                            </p>
                            <p className='log-link'>
                                Forgotten your username or password?{" "}
                                <Link to='/verify'>
                                    <span className='login-link'>
                                        Click here
                                    </span>
                                </Link>
                            </p>
                            <Alert />
                            <div className='login-submit-container'>
                                <PostButton classes='small-btn' value='Login' />
                                {loading && (
                                    <img
                                        className='login-loading-spinner'
                                        src={spinner}
                                        alt='loading'
                                    />
                                )}
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </section>
    );
};

LoginForm.propTypes = {
    username: PropTypes.string.isRequired,
    password: PropTypes.string.isRequired,
    loading: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
    isLoggedIn: PropTypes.bool,
    usernameField: PropTypes.object.isRequired
};

export default LoginForm;
