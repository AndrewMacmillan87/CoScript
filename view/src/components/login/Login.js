import React, { useState, useEffect, useRef } from "react";
import { Redirect } from "react-router-dom";

import NavBar from "../NavBar";
import LoginForm from "./LoginForm";

// Redux
import { connect } from "react-redux";
import { loginUser } from "../../redux/actions/auth";

import PropTypes from "prop-types";

import "./login.css";

export const Login = ({ ...props }) => {
    /**
     * Create a reference to the username field to focus
     * on it when the component mounts
     */
    const usernameRef = useRef();

    /**
     * Set state for this component and its children
     */
    const [loading, setLoading] = useState(false);
    const [formData, setFormData] = useState({
        username: "",
        password: "",
    });

    /**
     * Sets the browser tab title and scrolls to the top
     * of the page whe the component mounts. Also focuses
     * on the 'username' field
     */
    useEffect(() => {
        document.title = "CoScript - Login";
        window.scrollTo(0, 0);
        if (usernameRef.current) usernameRef.current.focus();
    }, []);

    /**
     * Sets the 'loading' state property to true in order
     * to display a loading spinner while checking login
     * credentials
     */
    const showSubmitSpinner = () => {
        setLoading(true);
    };

    /**
     * Sets the 'loading' state property to false in order
     * to remove a loading spinner if login credentials are
     * incorrect
     */
    const removeSubmitSpinner = () => {
        setLoading(false);
    };

    /**
     * Sets the state property <[e.target.name]> to the value
     * of the event target
     *
     * @param {Object} e - event object
     */
    const onChange = (e) => {
        setFormData({
            ...formData,
            [e.target.name]: e.target.value,
        });
    };

    /**
     * Attempts to log the user into the application. Shows
     * a loading spinner while validating credentials and removes
     * it if the credentials are invalid
     *
     * @param {Object} e -event object
     */
    const onSubmit = async (e) => {
        e.preventDefault();
        showSubmitSpinner();
        const success = await props.loginUser(
            formData.username,
            formData.password
        );
        if (!success) {
            removeSubmitSpinner();
        }
    };

    /**
     * Check the location the user is trying to navigate to
     * where they navigated from. Redirect the user if they
     * are not registered or are trying to access a protected
     * route
     */
    const { user, location, projectData } = props;

    const { from } = (location && location.state) || {
        from: { route: "/" },
    };

    if (user && user.isLoggedIn) {
        if (from.pathname && !projectData) {
            return from.pathname === "/update-credentials" ? (
                <Redirect to='/' />
            ) : (
                <Redirect to={from} />
            );
        }
        return <Redirect to={from} />;
    }

    return (
        <main>
            <NavBar />
            <LoginForm
                username={formData.username}
                password={formData.password}
                loading={loading}
                onChange={onChange}
                onSubmit={onSubmit}
                isLoggedIn={user ? user.isLoggedIn : false}
                usernameField={usernameRef}
            />
        </main>
    );
};

Login.propTypes = {
    user: PropTypes.object,
    projectData: PropTypes.object,
    loginUser: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
    projectData: state.documents.projectData,
});

export default connect(setProps, { loginUser })(Login);
