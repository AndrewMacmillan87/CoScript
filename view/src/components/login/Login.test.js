import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Login } from "./Login";
import LoginForm from "./LoginForm";
import NavBar from "../NavBar";
import PostButton from "../buttons/PostButton";

describe("Login component", () => {
    let props, mockLoginUser;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockLoginUser = jest.fn();
        props = {
            user: {},
            projectData: {},
            loginUser: mockLoginUser,
        };
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Login {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render the NavBar component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(NavBar).length).toEqual(1);
    });

    it("should render the LoginForm component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(LoginForm).length).toEqual(1);
    });

    it("should render the PostButton component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(PostButton).length).toEqual(1);
    });

    it("calls `loginUser()` when onSubmit is called", () => {
        jest.spyOn(Login.prototype, "onSubmit").mockImplementation(function () {
            mockLoginUser();
        });
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "username", value: "test101" },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "password", value: "password1234" },
            });
        wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });

        expect(mockLoginUser).toBeCalled();
    });
});

describe("LoginForm component", () => {
    let props, mockLoginUser;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockLoginUser = jest.fn();
        props = {
            user: {},
            projectData: {},
            loginUser: mockLoginUser,
        };
    });

    it("should render a section element with a className of login-cover", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(wrapper.find({ className: "login-cover" }).length).toEqual(1);
    });

    it("should render a h1 element with a className of login-header", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("h1").length).toEqual(1);
        expect(wrapper.find({ className: "login-header" }).length).toEqual(1);
    });

    it("should render a from element with a className of login-form", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("form").length).toEqual(1);
        expect(wrapper.find({ className: "login-form" }).length).toEqual(1);
    });

    it("should render 3 input elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("input").length).toEqual(3);
    });

    it("should render 2 label elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("label").length).toEqual(2);
    });

    it("should render 2 p elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(2);
    });

    it("should render 2 Link components", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Login {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Link).length).toEqual(2);
    });
});
