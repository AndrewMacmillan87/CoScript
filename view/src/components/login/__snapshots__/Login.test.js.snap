// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders correctly 1`] = `
<main>
  <nav
    className="navbar"
  />
  <section
    className="login-cover"
  >
    <div
      className="login-container"
    >
      <h1
        className="login-header"
      >
        Login...
      </h1>
      <div>
        <div
          className="login-form"
        >
          <form
            onSubmit={[Function]}
          >
            <label
              className="log-label"
              htmlFor="username"
            >
              Username:
            </label>
            <input
              className="log-input"
              name="username"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Username..."
              type="text"
              value=""
            />
            <br />
            <label
              className="log-label"
              htmlFor="password"
            >
              Password:
            </label>
            <input
              className="log-input"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password..."
              type="password"
              value=""
            />
            <br />
            <p
              className="log-link"
            >
              Haven't registered? You can do that
               
              <a
                href="/register"
                onClick={[Function]}
              >
                <span
                  className="login-link"
                >
                  here
                </span>
              </a>
            </p>
            <p
              className="log-link"
            >
              Forgotten your username or password?
               
              <a
                href="/verify"
                onClick={[Function]}
              >
                <span
                  className="login-link"
                >
                  Click here
                </span>
              </a>
            </p>
            <div
              className="login-submit-container"
            >
              <input
                className="small-btn"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</main>
`;
